transport-native-epoll/src/main/c/netty_epoll_linuxsocket.c

```diff
@@ -183,39 +183,39 @@ static void netty_epoll_linuxsocket_getTcpInfo(JNIEnv* env, jclass clazz, jint f
      if (netty_unix_socket_getOption(env, fd, IPPROTO_TCP, TCP_INFO, &tcp_info, sizeof(tcp_info)) == -1) {
          return;
      }
-     unsigned int cArray[32];
-     cArray[0] = tcp_info.tcpi_state;
-     cArray[1] = tcp_info.tcpi_ca_state;
-     cArray[2] = tcp_info.tcpi_retransmits;
-     cArray[3] = tcp_info.tcpi_probes;
-     cArray[4] = tcp_info.tcpi_backoff;
-     cArray[5] = tcp_info.tcpi_options;
-     cArray[6] = tcp_info.tcpi_snd_wscale;
-     cArray[7] = tcp_info.tcpi_rcv_wscale;
-     cArray[8] = tcp_info.tcpi_rto;
-     cArray[9] = tcp_info.tcpi_ato;
-     cArray[10] = tcp_info.tcpi_snd_mss;
-     cArray[11] = tcp_info.tcpi_rcv_mss;
-     cArray[12] = tcp_info.tcpi_unacked;
-     cArray[13] = tcp_info.tcpi_sacked;
-     cArray[14] = tcp_info.tcpi_lost;
-     cArray[15] = tcp_info.tcpi_retrans;
-     cArray[16] = tcp_info.tcpi_fackets;
-     cArray[17] = tcp_info.tcpi_last_data_sent;
-     cArray[18] = tcp_info.tcpi_last_ack_sent;
-     cArray[19] = tcp_info.tcpi_last_data_recv;
-     cArray[20] = tcp_info.tcpi_last_ack_recv;
-     cArray[21] = tcp_info.tcpi_pmtu;
-     cArray[22] = tcp_info.tcpi_rcv_ssthresh;
-     cArray[23] = tcp_info.tcpi_rtt;
-     cArray[24] = tcp_info.tcpi_rttvar;
-     cArray[25] = tcp_info.tcpi_snd_ssthresh;
-     cArray[26] = tcp_info.tcpi_snd_cwnd;
-     cArray[27] = tcp_info.tcpi_advmss;
-     cArray[28] = tcp_info.tcpi_reordering;
-     cArray[29] = tcp_info.tcpi_rcv_rtt;
-     cArray[30] = tcp_info.tcpi_rcv_space;
-     cArray[31] = tcp_info.tcpi_total_retrans;
+     jint cArray[32];
+     cArray[0] = (jint) tcp_info.tcpi_state;
+     cArray[1] = (jint) tcp_info.tcpi_ca_state;
+     cArray[2] = (jint) tcp_info.tcpi_retransmits;
+     cArray[3] = (jint) tcp_info.tcpi_probes;
+     cArray[4] = (jint) tcp_info.tcpi_backoff;
+     cArray[5] = (jint) tcp_info.tcpi_options;
+     cArray[6] = (jint) tcp_info.tcpi_snd_wscale;
+     cArray[7] = (jint) tcp_info.tcpi_rcv_wscale;
+     cArray[8] = (jint) tcp_info.tcpi_rto;
+     cArray[9] = (jint) tcp_info.tcpi_ato;
+     cArray[10] = (jint) tcp_info.tcpi_snd_mss;
+     cArray[11] = (jint) tcp_info.tcpi_rcv_mss;
+     cArray[12] = (jint) tcp_info.tcpi_unacked;
+     cArray[13] = (jint) tcp_info.tcpi_sacked;
+     cArray[14] = (jint) tcp_info.tcpi_lost;
+     cArray[15] = (jint) tcp_info.tcpi_retrans;
+     cArray[16] = (jint) tcp_info.tcpi_fackets;
+     cArray[17] = (jint) tcp_info.tcpi_last_data_sent;
+     cArray[18] = (jint) tcp_info.tcpi_last_ack_sent;
+     cArray[19] = (jint) tcp_info.tcpi_last_data_recv;
+     cArray[20] = (jint) tcp_info.tcpi_last_ack_recv;
```

Comment:
fwiw, casting them to int is not truncating them it just treats them as negative.  I view this as a java shortcoming and not something readily fixable.  At Google we use the Guava classes [UnsignedInts](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/primitives/UnsignedInts.html) and [UnsignedInteger](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/primitives/UnsignedInteger.html) to express that a number may appear negative.  

Using Guava is probably not possible for you, but it does mean consumers of Netty APIs have a way of interpretting the signed ints that come from tcp info.

Also, I don't think using longs really solves this.  Consumers still have to do bounds checking on the result, and arithmetic with longs and preexisting ints is more confusing than staying entirely with ints.